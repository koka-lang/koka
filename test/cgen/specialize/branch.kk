fun map_other(xs : list<a>, f : a -> b, g : a -> b) : list<b> {
    val isEven : bool = xs.length % 2 == 0
    match (xs) {
        Nil -> Nil
        // this branch gets specialized
        Cons(x, xx) | isEven    -> Cons(f(x), xx.map_other(f, g))
        // this branch does not since f and g are passed in a different order
                    | _         -> Cons(g(x), xx.map_other(g, f))
    }
}

fun main() {
    val plusone = fn(x) { x + 1 }
    val minusone = fn(x) { x - 1 }
    [1,2,3,4].map_other(plusone, minusone).foreach(println)
}