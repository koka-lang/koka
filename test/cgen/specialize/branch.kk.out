2
1
2
5
 
cgen/specialize/branch/.lift000-main: (xs : list<int>) -> console ()
cgen/specialize/branch/.lift000-main: forall<_e> (xs0 : list<int>) -> list<int>
cgen/specialize/branch/main: () -> console ()
cgen/specialize/branch/map_other: forall<a,b> (xs : list<a>, f : (a) -> b, g : (a) -> b) -> list<b>